src/System.Text.Json/src/System/Text/Json/Serialization/ReflectionEmitMemberAccessor.cs

```diff
@@ -0,0 +1,150 @@
+        }
+
+        public override Func<object, TProperty> CreatePropertyGetter<TClass, TProperty>(PropertyInfo propertyInfo) =>
+            (Func<object, TProperty>)CreatePropertyGetter(propertyInfo, typeof(TClass));
+
+        private static Delegate CreatePropertyGetter(PropertyInfo propertyInfo, Type classType)
+        {
+            MethodInfo realMethod = propertyInfo.GetGetMethod();
+            Type objectType = typeof(object);
+
+            var dynamicMethod = new DynamicMethod(
+                realMethod.Name,
+                propertyInfo.PropertyType,
+                new[] { objectType },
+                typeof(ReflectionEmitMemberAccessor).Module,
+                skipVisibility: true);
+
+            ILGenerator generator = dynamicMethod.GetILGenerator();
+
+            generator.Emit(OpCodes.Ldarg_0);
+
+            if (classType.IsValueType)
+            {
+                generator.Emit(OpCodes.Unbox_Any, classType);
+                generator.Emit(OpCodes.Call, realMethod);
+            }
+            else
+            {
+                generator.Emit(OpCodes.Castclass, classType);
+                generator.Emit(OpCodes.Callvirt, realMethod);
+            }
+
+            generator.Emit(OpCodes.Ret);
+
+            return dynamicMethod.CreateDelegate(typeof(Func<,>).MakeGenericType(objectType, propertyInfo.PropertyType));
+        }
+
+        public override Action<object, TProperty> CreatePropertySetter<TClass, TProperty>(PropertyInfo propertyInfo) =>
+            (Action<object, TProperty>)CreatePropertySetter(propertyInfo, typeof(TClass));
+
+        private static Delegate CreatePropertySetter(PropertyInfo propertyInfo, Type classType)
+        {
+            MethodInfo realMethod = propertyInfo.GetSetMethod();
+            Type objectType = typeof(object);
+
+            var dynamicMethod = new DynamicMethod(
+                realMethod.Name,
+                typeof(void),
+                new[] { objectType, propertyInfo.PropertyType },
+                typeof(ReflectionEmitMemberAccessor).Module,
+                skipVisibility: true);
+
+            ILGenerator generator = dynamicMethod.GetILGenerator();
+
+            generator.Emit(OpCodes.Ldarg_0);
+
+            if (classType.IsValueType)
+            {
+                generator.Emit(OpCodes.Unbox_Any, classType);
+                generator.Emit(OpCodes.Ldarg_1);
+                generator.Emit(OpCodes.Call, realMethod);
+            }
+            else
+            {
+                generator.Emit(OpCodes.Castclass, classType);
+                generator.Emit(OpCodes.Ldarg_1);
+                generator.Emit(OpCodes.Callvirt, realMethod);
+            };
+
+            generator.Emit(OpCodes.Ret);
+
+            return dynamicMethod.CreateDelegate(typeof(Action<,>).MakeGenericType(objectType, propertyInfo.PropertyType));
```

Comment:
For perforamcne, it is possible to remove `MakeGenericType` if this method takes `<TProperty>` generic and binds to a delegate on `JsonPropertyInfo<TProperty>` (or elsewhere).

Reply:
It's an initialization only method, so the performance isn't so critical here. At the same time if `TClass` or `TProperty` is a value type, the JIT will produce a separate codegens of it which will be used only once and will occupy memory.

With the current code the generic stub method `CreatePropertySetter<TClass, TProperty>` is 16 bytes long and (if my memory serves me right, @AndyAyersMS knows it) therefore it will be inlined. The jitted code of non-generic `CreatePropertySetter` is shared in that case.

Correct me if I'm wrong.

Reply:
While I was writing about optimizations I completely forgot that the method is virtual and cannot be inlined. I think with static interface members (dotnet/csharplang#1711) it would be possible to inline generics.

Reply:
Virtual methods can be inlined, if the jit can determine that they're called.

New generic instantiations always take up some space, whether or not their methods ever get called or inlined; the runtime has to keep track of them somehow. `Action` is relatively pay-for-play though as instantiating it does not trigger instantiation of other types.